server.js

```diff
@@ -7,23 +7,118 @@ const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts"
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+app.use(cors());
+app.use(bodyParser.json());
+
+// Middleware to handle server connection errors
+app.use((req, res, next) => {
+  try {
+    if (mongoose.connection.readyState === 1) {
+      next()
+    } else {
+      res.status(503).json({ error: 'Service unavailable' });
+    }
+  } catch (error) {
+    res.status(400).json({ error: 'Error! Could not access the server.' });
+  }
+});
+
+////////////// API ENDPOINTS ////////////// 
 
-// Start defining your routes here
+// To list all available endpoints on the starting page
+const listEndpoints = require('express-list-endpoints');
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  res.send(listEndpoints(app));
+});
+
+// Endpoint created to get a list of the 20 latest thoughts
+app.get('/thoughts', async (req, res) => {
+  // Field for client to sort on parameter
+  const sortField = req.query.sortField;
+  // Field for client to choose sorting order, 'asc' or 'desc'
+  const sortOrder = req.query.sortOrder;
+  // Creates the initial database query
+  let databaseQuery = Thought.find();
+
+  // To search with sortField and sortOrder with queries: 
+  // (GET '/thoughts?sortField=${sortField}&sortOrder=${sortOrder}')
+  if (sortField) {
+    databaseQuery = databaseQuery.sort({ [sortField]: sortOrder === 'asc' ? 1 : -1 })
+  } else {
+    databaseQuery = databaseQuery.sort({ createdAt: 'desc' }).limit(20);
+  }
+
+  // const thoughts = await Thought.find().sort({ createdAt: 'desc' }).limit(20).exec();
+
+  // The await function comes here instead after sorting / filtering with queries
+  const thoughts = await databaseQuery.exec();
+  res.status(200).json(thoughts);
+});
+
+// Endpoint for POSTing thoughts and adding it to the database
+app.post('/thoughts', async (req, res) => {
+  try {
+    // Success case
+    // Retrieves the information sent by the client to the API endpoint
+    const { message, name } = req.body;
+
+    // Use the mongoose model to create the database entry
+    const newThought = await new Thought({ message, name }).save();
+
+    res.status(200).json(newThought);
+  } catch (err) {
+    // Bad request
+    res.status(400).json({
+      message: 'Could not save thought to the Database',
+      errors: err.errors
+    });
+  }
+});
+
+// Endpoint for POSTing likes to thoughts in the list, recognize thought by ID
+app.post('/thoughts/:thoughtId/like', async (req, res) => {
+  try {
+    // Success case
+    await Thought.updateOne({ _id: req.params.thoughtId }, { $inc: { hearts: 1 } });
+    res.status(200).json();
+  }
+  catch (err) {
+    // Bad request
+    res.status(400).json({
+      message: 'Could not like this thought since it was not found.',
+      errors: err.errors
+    });
+  }
+});
 
 // Start the server
 app.listen(port, () => {
   console.log(`Server running on http://localhost:${port}`)
-})
+});
```

Comment:
Well written, clean code😃 I also like the comments that explains what all things do.